server/src/test/java/org/elasticsearch/search/aggregations/bucket/geogrid/GeoGridAggregatorTestBase.java

```diff
@@ -0,0 +1,147 @@
+import org.apache.lucene.index.DirectoryReader;
+import org.apache.lucene.index.IndexReader;
+import org.apache.lucene.index.RandomIndexWriter;
+import org.apache.lucene.search.IndexSearcher;
+import org.apache.lucene.search.MatchAllDocsQuery;
+import org.apache.lucene.search.Query;
+import org.apache.lucene.store.Directory;
+import org.elasticsearch.common.CheckedConsumer;
+import org.elasticsearch.index.mapper.GeoPointFieldMapper;
+import org.elasticsearch.index.mapper.MappedFieldType;
+import org.elasticsearch.search.aggregations.Aggregator;
+import org.elasticsearch.search.aggregations.AggregatorTestCase;
+import org.elasticsearch.search.aggregations.support.AggregationInspectionHelper;
+
+import java.io.IOException;
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+import java.util.function.Consumer;
+
+public abstract class GeoGridAggregatorTestBase<T extends InternalGeoGridBucket> extends AggregatorTestCase {
+
+    private static final String FIELD_NAME = "location";
+    private int minPrecision;
+
+    public GeoGridAggregatorTestBase(int minPrecision) {
+        this.minPrecision = minPrecision;
+    }
+
+    protected abstract int randomPrecision();
+
+    protected abstract String hashAsString(double lng, double lat, int precision);
+
+    protected double adjustLatitude(double lat) {
+        return lat;
+    }
+
+    protected double adjustLongitude(double lon) {
+        return lon;
+    }
+
+    public void testNoDocs() throws IOException {
+        testCase(new MatchAllDocsQuery(), FIELD_NAME, minPrecision, iw -> {
+            // Intentionally not writing any docs
+        }, geoGrid -> {
+            assertEquals(0, geoGrid.getBuckets().size());
+        });
+    }
+
+    public void testFieldMissing() throws IOException {
+        testCase(new MatchAllDocsQuery(), "wrong_field", minPrecision, iw -> {
+            iw.addDocument(Collections.singleton(new LatLonDocValuesField(FIELD_NAME, 10D, 10D)));
+        }, geoGrid -> {
+            assertEquals(0, geoGrid.getBuckets().size());
+        });
+    }
+
+    public void testWithSeveralDocs() throws IOException {
+        int precision = randomPrecision();
+        int numPoints = randomIntBetween(8, 128);
+        Map<String, Integer> expectedCountPerGeoHash = new HashMap<>();
+        testCase(new MatchAllDocsQuery(), FIELD_NAME, precision, iw -> {
+            List<LatLonDocValuesField> points = new ArrayList<>();
+            Set<String> distinctHashesPerDoc = new HashSet<>();
+            for (int pointId = 0; pointId < numPoints; pointId++) {
+                double lat = (180d * randomDouble()) - 90d;
+                double lng = (360d * randomDouble()) - 180d;
+
+                // Precision-adjust longitude/latitude to avoid wrong bucket placement
+                lng = adjustLongitude(lng);
```

Comment:
would you mind explaining this part? I am not sure I understand

Reply:
adustments have been removed for now - I will re-add them in a follow up PR with a detailed explanation.